File: bench/05_dataloader.py
import torch
import time
import os

batch_size = int(os.environ.get('BATCH_SIZE', "1000"))
device = os.environ.get('DEVICE', "cpu")

class Dataset (torch.utils.data.Dataset):
  def __init__ (self, ir, batch_size):
    self.l = ir*batch_size
    self.x = torch.randn(ir*batch_size, 28, 28)
    self.y = torch.randint(10, (ir*batch_size,))
  def __getitem__ (self, i):
    return (self.x[i], self.y[i])
  def __len__ (self):
    return self.l

if device == "cpu":
  def fn ():
    for el in torch.utils.data.DataLoader(ds, batch_size = batch_size):
      x = el[0]
      y = el[1]
else:
  def fn():
    for el in torch.utils.data.DataLoader(ds, batch_size = batch_size):
      x = el[0].cuda()
      y = el[1].cuda()
    torch.cuda.synchronize()

ir = 1
ds = Dataset(ir, batch_size)

fn()

ir = int(os.environ.get('ITER', "1000"))
ds = Dataset(ir, batch_size)

start_time = time.time()
fn()
print((time.time() - start_time))

File: bench/05_dataloader.R
library(torch)

batch_size <- as.numeric(Sys.getenv("BATCH_SIZE", unset = "1000"))
device <- Sys.getenv("DEVICE", unset = "cpu")

Dataset <- dataset(
  initialize = function(ir, batch_size) {
    self$len <- ir*batch_size
    self$x <- torch_randn(self$len, 28, 28)
    self$y <- torch_randint(1, 10, size = c(self$len))
  },
  .getbatch = function(i) {
    list(
      self$x[i],
      self$y[i]
    )
  },
  .length = function() {
    self$len
  }
)

if (device == "cpu") {
  fn <- function() {
    coro::loop(for(el in dataloader(ds, batch_size = batch_size)) {
      x <- el[[1]]
      y <- el[[2]]
    })
    invisible(NULL)
  }
} else {
  fn <- function() {
    coro::loop(for(el in dataloader(ds, batch_size = batch_size)) {
      x <- el[[1]]$cuda()
      y <- el[[2]]$cuda()
    })
    cuda_synchronize()
    invisible(NULL)
  }
}

iter <- 1
ds <- Dataset(iter, batch_size)

fn()

iter <- as.numeric(Sys.getenv("ITER", unset = "100"))
ds <- Dataset(iter, batch_size)

cat(system.time(fn())[["elapsed"]])
